File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.time.LocalDate;
import java.util.HashMap;
import java.util.ArrayList;


class Solution {
    private int getDate(String date){
        String[] parts = date.split("\\.");
        int year = Integer.parseInt(parts[0]);
        int month = Integer.parseInt(parts[1]);
        int day = Integer.parseInt(parts[2]);
        return year * 12 * 28 + month * 28 + day;
    }
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        int todayDate = getDate(today);
        
        HashMap<String, Integer> termMap = new HashMap<>();
        for(String term: terms){
            String[] parts = term.split(" ");
            termMap.put(parts[0], Integer.parseInt(parts[1]));
        }
        
        ArrayList<Integer> answer = new ArrayList<>();
        for(int i=0; i < privacies.length; i++){
            String[] parts = privacies[i].split(" ");
            
            if(getDate(parts[0]) + termMap.get(parts[1]) * 28 - 1 < todayDate){
                answer.add(i+1);
            }
        }

        return answer.stream().mapToInt(integer -> integer).toArray();
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    t_year, t_month, t_day = map(int, today.split('.'))
    today_c = t_year * 12 * 28 + t_month * 28 + t_day
    
    term_dict = dict()
    for term in terms:
        privacy_type, valid_duration = term.split()
        term_dict[privacy_type] = int(valid_duration)
    
    
    answer = []
    
    for i, privacy in enumerate(privacies):
        date, privacy_type = privacy.split()
        year, month, day = map(int, date.split('.'))
        c = year * 12 * 28 + month * 28 + day
        e_c = c + term_dict[privacy_type] * 28
        
        if today_c > e_c - 1:
            answer.append(i+1)
        
    return answer